### ISSUE TITLE
resource_dns_mx_record_set.go:resourceDnsMXRecordSetHash panics with apply-time string interpolation

### ISSUE THREAD

**Issue by diekmann** [hashicorp/terraform-provider-dns#82]
### Terraform Version
Terraform v0.12.0-beta1
+ provider.dns v2.1.1

### Affected Resource(s)
- dns_mx_record_set

### Terraform Configuration Files
```hcl
provider "dns" {
  update {
    server    = "127.0.0.1" # nothing running there
  }
}

resource "dns_mx_record_set" "mx" {
  zone = "example.com."

  dynamic "mx" {
    for_each = split(" ", "1 ${dns_mx_record_set.mx.id}")
    content {
      preference = tonumber(mx.value)
      exchange   = "foo${mx.value}.example.com."
    }
  }
}
```

With the self-referencing `for_each` statement, I force the evaluation of the `preference` and `exchange` to apply-time. There `resourceDnsMXRecordSetHash()` crashes because it gets passed a `nil`.

### Output
```text
$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # dns_mx_record_set.mx will be created
  + resource "dns_mx_record_set" "mx" {
      + id   = (known after apply)
      + ttl  = 3600
      + zone = "example.com."

      + mx {
          + exchange   = (known after apply)
          + preference = (known after apply)
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

dns_mx_record_set.mx: Creating...

Error: rpc error: code = Unavailable desc = transport is closing


panic: interface conversion: interface {} is nil, not map[string]interface {}
2019-05-06T12:30:50.275Z [DEBUG] plugin.terraform-provider-dns_v2.1.1_x4: 
2019-05-06T12:30:50.275Z [DEBUG] plugin.terraform-provider-dns_v2.1.1_x4: goroutine 22 [running]:
2019-05-06T12:30:50.275Z [DEBUG] plugin.terraform-provider-dns_v2.1.1_x4: github.com/terraform-providers/terraform-provider-dns/dns.resourceDnsMXRecordSetHash(0x0, 0x0, 0x1)
2019-05-06T12:30:50.275Z [DEBUG] plugin.terraform-provider-dns_v2.1.1_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-dns/dns/resource_dns_mx_record_set.go:162 +0x312
2019-05-06T12:30:50.275Z [DEBUG] plugin.terraform-provider-dns_v2.1.1_x4: github.com/terraform-providers/terraform-provider-dns/vendor/github.com/hashicorp/terraform/helper/schema.(*Set).hash(0xc0002559e0, 0x0, 0x0, 0x3, 0xc0000f7600)
...
```

### Panic Output
`crash.log`: https://gist.github.com/diekmann/fc2ef2bf41ac3bd0524b2a8075ae5124

### Expected Behavior
Not sure, but no crash :-)

### Actual Behavior
Crash

### Steps to Reproduce
1. `terraform apply`

**Comment by bodgit** [comment]
I'm struggling to work out what you're trying to do here. It looks like you're using the ID of the resource within itself. How will that work?

**Comment by diekmann** [comment]
Indeed, my use case in 100% contrived. Also, Terraform v0.12.0-beta2 rejects this self reference.

The provided config is the simplest example I could come up with to trigger the crash.

My actual bug report is: If all values of [`mx`](https://www.terraform.io/docs/providers/dns/r/dns_mx_record_set.html#mx-1) are only `(known after apply)`, `resourceDnsMXRecordSetHash()` crashes.

If you think there is absolutely no way that a reasonable use case can ever trigger this scenario, feel free to close this bug 😸

**Comment by bodgit** [comment]
Right, I see what you mean now. I don't think there's ever a valid case for _all_ values of `mx` to be `(known after apply)` as the record cannot be created at all. Not sure how/where that should be enforced, if it's something that needs to be added to that resource, (and likely the other resources that also have "complex" values, SRV for example), or enforced further up.

**Comment by apparentlymart** [comment]
Thanks for reporting this, @diekmann!

Indeed you are right that this ought not to crash. I think the reason it is crashing is a disagreement between the older SDK code (which is written in terms of the Terraform 0.11 provider protocol) and the translation layer we are using to adapt the existing providers to Terraform 0.12's protocol.

In particular, it seems that this wholly-unknown block is being converted into the old model in an unexpected way, and thus it's ending up in the set has function as `nil` instead of as an object. We made some changes to the SDK translation layer recently which are not yet reflected in a DNS provider release, so it's possible that just upgrading that provider's version of the SDK will make this go away.

If not, we probably ought to have an additional check to ensure that the value passed to the set hash function can never end up being `nil`, but if this is the only provider encountering the problem (we'll see what else comes up during testing) we may instead choose to fix it at the provider level just to keep the fix localized, and thus avoid the risk of regressions for other providers.

**Comment by bflad** [comment]
The `dns_mx_record_set` managed resource code was migrated to terraform-plugin-framework and released in version 3.3.0. Any resource crashes with provider versions after 3.3.0 should be filed as new bug reports since the code is wholly different now.
